Monorepo Support

InTheCloudDan opened the pull request:
Initial monorepo support. Mostly implemented in a backwards compatible way. Anyone running via CLI args today are supported.

Added:
* `project` block in `coderefs.yaml`. You can now supply a list of multiple projects, including their starting subdirectory and additional alias configurations specific to that project.
* Check so that `projKey` and new `projects` blocks cannot be used at the same time.

Changes:
* `accessToken` is used for all projects.
* `aliases` block is now global. If an alias is added there it will affect all projects.

Breaking Changes:
* When using `--outDir` we previously included the LaunchDarkly Project name in output file name, this has been removed because it may contain flags from multiple projects.

Testing:
Add a `coderefs.yaml` under `.launchdarkly` directory that contains multiple projects and optional dir along with aliases.
```
projects:
  - projectKey: code-refs-test
    dir: public/
    aliases:
       - type: camelcase
  - projectKey: demo-test-project

```

InTheCloudDan commented on coderefs/coderefs.go:
```diff
@@ -15,14 +15,19 @@ import (
 )
 
 func Run(opts options.Options) {
+	if len(opts.ProjKey) > 0 {
```
We set this inside run because all instances(github actions, bitbucket, CLI) call to `Run`. Converts legacy `projKey` into a `Project`.

InTheCloudDan commented on coderefs/coderefs.go:
```diff
@@ -15,14 +15,19 @@ import (
 
 	log.Info.Printf("absolute directory path: %s", absPath)
-	ldApi := ld.InitApiClient(ld.ApiOptions{ApiKey: opts.AccessToken, BaseUri: opts.BaseUri, ProjKey: opts.ProjKey, UserAgent: "LDFindCodeRefs/" + version.Version})
+	ldApi := ld.InitApiClient(ld.ApiOptions{ApiKey: opts.AccessToken, BaseUri: opts.BaseUri, UserAgent: "LDFindCodeRefs/" + version.Version})
```
Only used to get the project for flags. Changed the calling function to pass in the relevant project.

InTheCloudDan commented on coderefs/coderefs.go:
```diff
@@ -108,8 +113,13 @@ func calculateStaleBranches(branches []ld.BranchRep, remoteBranches map[string]b
+
 	if outDir != "" {
-		outPath, err := branch.WriteToCSV(outDir, opts.ProjKey, repoParams.Name, opts.Revision)
+		outPath, err := branch.WriteToCSV(outDir, repoParams.Name, opts.Revision)
```
As noted in PR description. We no longer include the `projKey` in the output file name so this is a breaking change.

InTheCloudDan commented on coderefs/coderefs.go:
```diff
@@ -154,30 +168,31 @@ func runExtinctions(opts options.Options, matcher search.Matcher, branch ld.Bran
-				missingFlags = append(missingFlags, flag)
-			}
+		var removedFlags []ld.ExtinctionRep
+		for i, project := range opts.Projects {
```
Loop over all of the projects now to find extinctions. The API endpoint already accepts an array of extinctions and each extinction has its project key embedded in the struct.

InTheCloudDan commented on flags/flags.go:
```diff
@@ -15,35 +15,37 @@ const (
-	projKey := opts.ProjKey
+func GetFlagKeys(opts options.Options, repoParams ld.RepoParams) map[string][]string {
+	flagKeys := make(map[string][]string)
+	for _, proj := range opts.Projects {
```
Performing the loop in here because it already is coupled pretty tightly to implementation. Down the line we could look to move this outside of the function.

InTheCloudDan commented on internal/git/git.go:
```diff
@@ -183,40 +185,49 @@ func (c Client) FindExtinctions(projKey string, flags []string, matcher search.M
 			return nil, err
 		}
-		patch, err := changes.Patch()
+		patch, err := changes.PatchContext(context.Background())
```
This allows us to access `FilePatches` below. So now we can loop over the files in the commit to see if they're in the subdirectory for that matcher.

InTheCloudDan commented on internal/git/git.go:
```diff
@@ -183,40 +185,49 @@ func (c Client) FindExtinctions(projKey string, flags []string, matcher search.M
-					removalCount += delta
+		for _, filePatch := range patch.FilePatches() {
+			fromFile, toFile := filePatch.Files()
+			if project.Dir != "" && (toFile == nil || !strings.HasPrefix(toFile.Path(), project.Dir)) {
```
Tested this logic with local changes. Need to add more test coverage, also would like to talk over the implementation.

InTheCloudDan replied:
We added validation logic in a separate PR building on this one that the `project.Dir` has to be relative. Given that and how go-git returns paths as relative to the root of the repo this and the conditional inside it are checking if the path does not start the `projects` `dir`. If not then continue onto next patch

InTheCloudDan commented on internal/ld/ld.go:
```diff
@@ -91,10 +91,10 @@ func InitApiClient(options ApiOptions) ApiClient {
 }
 
-func (c ApiClient) GetFlagKeyList() ([]string, error) {
+func (c ApiClient) GetFlagKeyList(projKey string) ([]string, error) {
```
Mentioned above. Now we explicitly pass in the project instead of relying on the `options.ProjKey`

InTheCloudDan commented on internal/ld/ld.go:
```diff
@@ -551,23 +551,27 @@ func (t tableData) Swap(i, j int) {
 	for _, ref := range b.References {
 		for _, hunk := range ref.Hunks {
-			refCountByFlag[hunk.FlagKey]++
+			if hunk.ProjKey == project || project == "" {
```
Allowing empty `""` for the testing below.

InTheCloudDan commented on coderefs/coderefs.go:
```diff
@@ -154,30 +168,31 @@ func runExtinctions(opts options.Options, matcher search.Matcher, branch ld.Bran
+		var removedFlags []ld.ExtinctionRep
+		for i, project := range opts.Projects {
+			missingFlags := []string{}
+			for flag, count := range branch.CountByFlag(matcher.Elements[i].Elements, project.Key) {
```
For how we generated and assign the matcher elements, there is one for each Project struct so we can use the index `i` here to look the corresponding values.

InTheCloudDan commented on coderefs/coderefs.go:
```diff
@@ -121,21 +131,25 @@ func handleOutput(opts options.Options, matcher search.Matcher, branch ld.Branch
 	if opts.DryRun {
+		totalFlags := 0
+		for _, searchElems := range matcher.Elements {
+			totalFlags += len(searchElems.Elements)
```
Verify elements is correct number of flags.

InTheCloudDan commented on flags/flags.go:
```diff
@@ -15,35 +15,37 @@ const (
-	isDryRun := opts.DryRun
+		ignoreServiceErrors := opts.IgnoreServiceErrors
+		if !isDryRun {
+			err := ldApi.MaybeUpsertCodeReferenceRepository(repoParams)
```
Can this be moved outside of project loop.

InTheCloudDan replied:
fixed

InTheCloudDan commented on options/options.go:
```diff
@@ -165,11 +171,20 @@ func (o Options) ValidateRequired() error {
+		return fmt.Errorf("`--projKey` cannot be combined with `projects` in configuration")
+	}
+
+	if len(o.ProjKey) > maxProjKeyLength && len(o.Projects) == 0 {
```
remove `&& len(o.Projects) == 0` if `o.ProjKey` is less than maxProjKeyLength it shouldn't matter if projects is not empty.

yusinto commented:
This looks good! Apart from the dir ~ relative path issue we discussed in the [blitz](https://launchdarkly.atlassian.net/wiki/spaces/ENG/pages/1985446495/Monorepo+Blitz+-+Week+of+12+06) everything else seems to work!

InTheCloudDan commented:
@XieX I updated the sections of code we talked over. I also added a few new helper functions. Rather than looping over all of the references for each project in a loop multiple times we now return a `map[string]map[string]int64` which is `[project][flag]int` so it's less processing.

jazanne reviewed (approved):
i'm so excited about this change!
